### PR TITLE
don't block istiod readiness on remote clusters

### PR DESCRIPTION
**Pull request by stevenctl**
alternative to #31768 

We want to avoid starting a new istiod instance when it can't reach remote clusters due to transient errors.
We also want to avoid persistent errors connecting to other clusters from stopping any istiod's from coming up. 

To compromise, components that read from remote clusters should expose both: `HasSynced` and an additional function `SyncErr`

While waiting for remote cluster controllers to sync, we will block istiod's readiness. `SyncErr` lets us know if `HasSynced` is returning `false` due to an error rather than us just taking a while to sync. 

After a timeout, if we find an error, we allow the server to ready, assuming the error is persistent.
If there is no error after the timeout, we assume the startup is slow. 

Istiod will _not_ ready if any controllers for the local/config cluster fail to flip `HasSynced` to `true`. 

---

This approach has tradeoffs: it leaves istiods that have endpoints for remote clusters whose API servers are no longer reachable, allowing new ones to come up with inconsistent endpoints _eventually_. 

Somehow we need to either get rid of the old istiods or have them drop the endpoints, if we consider API server reachability to be tied with the endpoints' health.

### CODE REVIEW

**Comment by stevenctl** [comment]
cc @costinm @howardjohn

**Comment by howardjohn** [comment]
> another alternative is to make a single attempt at making the other clusters ready, but if a single failure occurs we move on; this can be difficult because each registry marks HasReady async

I am concerned with not doing this option. A concrete example of where this would be a regression for users:
* `my-service`, in primary cluster calls out to `dependency` in cluster `remote`. ALL endpoints are in cluster `remote`
* New istiod comes up, and it takes a while to connect/sync to `remote cluster`.
* `my-service` pods connect to the new istiod instance, and get no endpoints for `dependency`.
* All calls now 503

Its tricky - really I would want to only fail-open if ALL istiod's cannot connect. That behavior is unlikely to be implementable though. However, we should at least ensure the error is not just a transient issues, and especially must not give this behavior just due to startup race conditions?

**Comment by stevenctl** [comment]
I really don't want to merge this PR (as-is). I just want to try to get something leaner than https://github.com/istio/istio/pull/31768. 

> another way to handle this is making our controllers include a SyncErrors in addition to HasSynced. With that, we could wait indefinitely when we expect to eventually succeed

We'd have to be somewhat intelligent in how we set SyncErrors != nil.

**Comment by stevenctl** [comment]
It's really hard to avoid the issue of having old istiods send the cross-cluster endpoints, and new ones not having them. The two directions are:

1. Don't let new istiods come up
2. Make the existing istiods purge endpoints when a cluster isn't reachable for long enough

1 isn't acceptable without some qualifiers. 
2 ties API server availability with Pod availability. 


We may need some combination of the 2. 

---

Going down path 1, we can say "don't let a new istiod come up if the errors seem transient". 

The rules for istiod readiness would be:
1. The local/config cluster registry HasSynced
2. Multicluster registries HasSynced OR (SyncError != nil AND Timeout expired)
 
We let new istiods come up _eventually_ to avoid the mesh from breaking down. 

If SyncErr is nil, we're just taking a long time to process everything, so the timeout shouldn't apply. We only use that to say, "errors have persisted for a while, this probably isn't going away".

**Comment by howardjohn** [comment]
> The local/config cluster registry HasSynced
> Multicluster registries HasSynced OR (SyncError != nil AND Timeout expired)

I think that is probably the best option

My one concern is a user leaves around a bad remote secret forever, and now all istiod pods are delayed by `Timeout`. That is probably ok, we can work around it with clear logs? If it becomes an issue, we *could* write into the Secret some annotation to disable it if its broken for more than X minutes, so other instances can fail-fast. But that is probably overkill for now.

File: pilot/pkg/config/kube/crdclient/client.go [hunk 1]
```diff
@@ -133,6 +135,17 @@ func (cl *Client) HasSynced() bool {
 	return true
 }
 
+func (cl *Client) SyncErr() error {
+	req, _ := klabels.NewRequirement("noop.istio.io/noop", "=", []string{"noop"})
+	var errs error
+	for _, ctl := range cl.kinds {
+		if _, err := ctl.lister("default").List(klabels.NewSelector().Add(*req)); err != nil {
```
Comment (stevenctl) [inline pilot/pkg/config/kube/crdclient/client.go:142]: the errors we see in `klog` happen somewhere inside the informers where we can't react or detect them..  resorting to calling listers unless we have a better idea

File: pilot/pkg/serviceregistry/kube/controller/multicluster.go [hunk 2]
```diff
@@ -377,8 +392,91 @@ func (m *Multicluster) InitSecretController(stop <-chan struct{}) {
...
+	cache.WaitForCacheSync(stop, m.secretController.HasSynced)
+}
+
+func (m *Multicluster) waitForSyncErrors(stop chan struct{}) error {
+	syncs := make([]func() error, 0, len(m.remoteKubeControllers))
+	m.m.Lock()
+	for _, controller := range m.remoteKubeControllers {
+		syncs = append(syncs, controller.SyncError)
+		if controller.configStore != nil {
+			syncs = append(syncs, controller.configStore.SyncErr)
```
Comment (stevenctl) [inline pilot/pkg/serviceregistry/kube/controller/multicluster.go:461]: The config store SyncErr is a lot more expensive – it calls a lister, whereas the kube controllers' SyncErr impl has the error precomputed based on what's going on in HasSynced...

Comment (hzxuzhonghu) [reply pilot/pkg/serviceregistry/kube/controller/multicluster.go:461]: What is `SyncErr` used for?

**Comment by hzxuzhonghu** [comment]
I am worried this will mask the real error and make the mesh inconsistent across different control plane instances.

**Comment by stevenctl** [comment]
@hzxuzhonghu it will definitely cause inconsistency – control planes that _were_ able to reach a cluster and can no longer reach it should either shut down or gracefully drop endpoints. That's the only way I see that we can get back in sync. 

I _don't_ think it masks the real error. Users will still have their healthy local cluster running, and to troubleshoot why their multicluster installation isn't working they should just check some logs. 

We could also expose some debug endpoints to list remote clusters and their health/reachability, and even add debugging tools to istiod.

**Comment by ramaraochavali** [comment]
> We could also expose some debug endpoints to list remote clusters and their health/reachability, and even add debugging tools to istiod.

I think `/clusters` end point that lists clusters and health status would be useful

**Comment by costinm** [comment]
> > The local/config cluster registry HasSynced
> > Multicluster registries HasSynced OR (SyncError != nil AND Timeout expired)
> 
> I think that is probably the best option
> 
> My one concern is a user leaves around a bad remote secret forever, and now all istiod pods are delayed by `Timeout`. That is probably ok, we can work around it with clear logs? If it becomes an issue, we _could_ write into the Secret some annotation to disable it if its broken for more than X minutes, so other instances can fail-fast. But that is probably overkill for now.

We do need to retry connecting to the 'syncErr != nil' - maybe every 15 min or so. 

And please expose this via metrics that we can alert on - logs are useless in large production systems, very hard to alert on logs and on-call doesn't spend their time watching logs..

**Comment by stevenctl** [comment]
> We do need to retry connecting to the 'syncErr != nil' - maybe every 15 min or so.

We will still indefinitely wait for `HasReady`. As soon as it passes, we will start to use the remote clusters. SyncErr is purely used to stop blocking our readiness probes.

File: pilot/pkg/config/kube/crdclient/client.go (hunk 1 above)
Comment (hzxuzhonghu) [reply pilot/pkg/config/kube/crdclient/client.go:142]: After a closer look at List method, i think it is not right to judge the access to remote clusters based on the error returned. In reality, there is no error even the client can not access remote cluster.

Comment (hzxuzhonghu) [reply pilot/pkg/config/kube/crdclient/client.go:141]: @stevenctl note this

**Pull request closed without merging**